### PR TITLE
Add a NonceCache interface to check nonces and prevent replay attacks.

### PR DESCRIPTION
**Pull request by jklein24**
This enforces that clients consider the nonce mechanisms involved in the protocol so it doesn't slip through the cracks (like it has in the demo vasp).

### CODE REVIEW

File: uma/nonce_cache.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	// This allows the cache to be pruned periodically while still preventing replay attacks by holding onto an earliest
+	// timestamp that is still valid.
+	PurgeNoncesOlderThan(timestamp time.Time)
+}
+
+// InMemoryNonceCache is an in-memory implementation of NonceCache.
+// It is not recommended to use this in production, as it will not persist across restarts. You likely want to implement
+// your own NonceCache that persists to a database of some sort.
+type InMemoryNonceCache struct {
+	cache                map[string]time.Time
```
Comment (mhrheaume) [inline uma/nonce_cache.go]: Use a `sync.Map` if this should be thread safe? Or since this is just used for testing maybe it's fine?

**Review by mhrheaume — COMMENTED** [review]
LGTM!

File: uma/nonce_cache.go (hunk 1 above)
Comment (jklein24) [reply uma/nonce_cache.go]: Thanks for the suggestion! Fixed.

File: uma/nonce_cache.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		cache:                sync.Map{},
+		oldestValidTimestamp: oldestValidTimestamp,
+	}
+}
+
+func (c *InMemoryNonceCache) CheckAndSaveNonce(nonce string, timestamp time.Time) error {
+	if _, ok := c.cache.Load(nonce); ok {
+		return errors.New("nonce already used")
+	}
+	if timestamp.Before(c.oldestValidTimestamp) {
```
Comment (mhrheaume) [inline uma/nonce_cache.go:39]: I think check this first (before doing anything with the map) and then do a `LoadOrStore` to atomically store the nonce (or error if it's already there).

Comment (jklein24) [reply uma/nonce_cache.go:39]: Good call. Done.

**Review by mhrheaume — APPROVED** [review]

**Pull request merged**